### PR TITLE
[SPARK-45762][CORE] Support shuffle managers defined in user jars by changing startup order

### PR DESCRIPTION
**Pull request by abellina**
### What changes were proposed in this pull request?
As reported here https://issues.apache.org/jira/browse/SPARK-45762, `ShuffleManager` instances defined in a user jar cannot be used in all cases, unless specified in the `extraClassPath`. We would like to avoid adding extra configurations if this instance is already included in a jar passed via `--jars`.

Proposed changes:

Refactor code so we initialize the `ShuffleManager` later, after jars have been localized. This is especially necessary in the executor, where we would need to move this initialization until after the `replClassLoader` is updated with jars passed in `--jars`.

Before this change, the `ShuffleManager` is instantiated at `SparkEnv` creation. Having to instantiate the `ShuffleManager` this early doesn't work, because user jars have not been localized in all scenarios, and we will fail to load the `ShuffleManager` defined in `--jars`. We propose moving the `ShuffleManager` instantiation to `SparkContext` on the driver, and `Executor`.

### Why are the changes needed?
This is not a new API but a change of startup order. The changed are needed to improve the user experience for the user by reducing extra configurations depending on how a spark application is launched.

### Does this PR introduce _any_ user-facing change?
Yes, but it's backwards compatible. Users no longer need to specify a `ShuffleManager` jar in `extraClassPath`, but they are able to if they desire.

This change is not binary compatible with Spark 3.5.0 (see MIMA comments below). I have added a rule to MimaExcludes to handle it https://github.com/apache/spark/pull/43627/commits/970bff4edc6ba14d8de78aa175415e204d6a627b

### How was this patch tested?
Added a unit test showing that a test `ShuffleManager` is available after `--jars` are passed, but not without (using local-cluster mode). 

Tested manually with standalone mode, local-cluster mode, yarn client and cluster mode, k8s.

### Was this patch authored or co-authored using generative AI tooling?
No

### CODE REVIEW

**Comment by abellina** [comment]
I'll have to figure out the CI. It seems my fork is running things, but I am getting some failures in this page (`AppVeyor` and the `Notify test workflow`)

**Comment by abellina** [comment]
The MIMA tests are failing due to:
```
[error] spark-core: Failed binary compatibility check against org.apache.spark:spark-core_2.13:3.5.0! Found 1 potential problems (filtered 3908)
[error]  * method this(java.lang.String,org.apache.spark.rpc.RpcEnv,org.apache.spark.serializer.Serializer,org.apache.spark.serializer.Serializer,org.apache.spark.serializer.SerializerManager,org.apache.spark.MapOutputTracker,org.apache.spark.shuffle.ShuffleManager,org.apache.spark.broadcast.BroadcastManager,org.apache.spark.storage.BlockManager,org.apache.spark.SecurityManager,org.apache.spark.metrics.MetricsSystem,org.apache.spark.memory.MemoryManager,org.apache.spark.scheduler.OutputCommitCoordinator,org.apache.spark.SparkConf)Unit in class org.apache.spark.SparkEnv does not have a correspondent in current version
```
Which makes sense, since I changed `SparkEnv`.

I am not entirely sure if adding this to `MimaExcludes` is the right approach here, and I think I need some help.

**Comment by abellina** [comment]
Ok I believe given the mima code that I need to add a temporary restriction here: https://github.com/apache/spark/blob/master/project/MimaExcludes.scala#L37.

**Comment by abellina** [comment]
@tgravescs fyi

File: core/src/main/scala/org/apache/spark/SparkEnv.scala [hunk 1]
```diff
@@ -71,6 +69,9 @@ class SparkEnv (
     val outputCommitCoordinator: OutputCommitCoordinator,
     val conf: SparkConf) extends Logging {
 
+  // We set the ShuffleManager in SparkContext and Executor
```
Comment (tgravescs) [inline core/src/main/scala/org/apache/spark/SparkEnv.scala:72]: nit update comment to say something like: the ShuffleManager is initialized later in... to allow it being defined in user specified jars.

File: core/src/test/scala/org/apache/spark/storage/BlockManagerReplicationSuite.scala [hunk 2]
```diff
@@ -104,8 +105,8 @@ trait BlockManagerReplicationBehavior extends SparkFunSuite
     val blockManagerInfo = new mutable.HashMap[BlockManagerId, BlockManagerInfo]()
     master = new BlockManagerMaster(rpcEnv.setupEndpoint("blockmanager",
       new BlockManagerMasterEndpoint(rpcEnv, true, conf,
-        new LiveListenerBus(conf), None, blockManagerInfo, mapOutputTracker, sc.env.shuffleManager,
-        isDriver = true)),
+        new LiveListenerBus(conf), None, blockManagerInfo, mapOutputTracker,
+        sc.env.shuffleManager.shuffleBlockResolver.getBlocksForShuffle, true)),
```
Comment (tgravescs) [inline core/src/test/scala/org/apache/spark/storage/BlockManagerReplicationSuite.scala:109]: put back the isDriver = true as last parameter

File: core/src/main/scala/org/apache/spark/storage/BlockManager.scala [hunk 3]
```diff
@@ -186,12 +186,14 @@ private[spark] class BlockManager(
...
     memoryManager: MemoryManager,
     mapOutputTracker: MapOutputTracker,
-    shuffleManager: ShuffleManager,
     val blockTransferService: BlockTransferService,
     securityManager: SecurityManager,
     externalBlockStoreClient: Option[ExternalBlockStoreClient])
   extends BlockDataManager with BlockEvictionHandler with Logging {
 
+  // this is set after the ShuffleManager is instantiated in SparkContext and Executor
+  private var shuffleManager: ShuffleManager = _
```
Comment (tgravescs) [inline core/src/main/scala/org/apache/spark/storage/BlockManager.scala:195]: update description above to mention having to set the shuffle manager as well.

File: core/src/main/scala/org/apache/spark/SparkEnv.scala [hunk 4]
```diff
@@ -423,7 +425,6 @@ object SparkEnv extends Logging {
       conf,
       memoryManager,
```
Comment (tgravescs) [inline core/src/main/scala/org/apache/spark/SparkEnv.scala:426]: there is a comment above that says "     // NB: blockManager is not valid until initialize() is called later."   We may want to update that to mention having to set shuffle manager

**Review by tgravescs — COMMENTED** [review]
Overall looks fine, it does complicate the initialization a bit but not sure I see a better way to handle that.   Would be good to get more eyes on it.

cc @mridulm  since I think you looked at shuffle related stuff in past.

**Comment by abellina** [comment]
@tgravescs thanks for the review. I have handled your comments in this commit: https://github.com/apache/spark/pull/43627/commits/0bd7e990812d23166509ad6585c8d352f78e569f

File: core/src/main/scala/org/apache/spark/SparkEnv.scala [hunk 5]
```diff
@@ -71,6 +69,10 @@ class SparkEnv (
     val outputCommitCoordinator: OutputCommitCoordinator,
     val conf: SparkConf) extends Logging {
 
+  // We initialize the ShuffleManager later, in SparkContext and Executor, to allow
+  // user jars to define custom ShuffleManagers.
+  var shuffleManager: ShuffleManager = _
```
Comment (mridulm) [inline core/src/main/scala/org/apache/spark/SparkEnv.scala:74]: Given `SparkEnv` is a `DeveloperApi`, let us not expose this for mutation.

```suggestion
  private var _shuffleManager: ShuffleManager = _
  
  def shuffleManager: ShuffleManager = _shuffleManager
```

**Review by mridulm — COMMENTED** [review]
Given there is a way for users to specify this reasonably right now, the amount of change to add support for this looks a bit lot.
Thoughts @tgravescs ?

File: core/src/main/scala/org/apache/spark/SparkEnv.scala (hunk 5 above)
Comment (abellina) [reply core/src/main/scala/org/apache/spark/SparkEnv.scala:74]: Will fix !!

**Comment by abellina** [comment]
@mridulm thanks for the comments. I have published a SPIP here https://issues.apache.org/jira/browse/SPARK-45792 that aims to show the bigger picture. Without the change of initialization order in this PR, we couldn't carry out the SPIP linked, because the ShuffleManager is initialized really early in the Executors today. I split this up into a separate PR to not introduce too much change at once, but your point is well taken. I would like to hear your thoughts around the SPIP and how we can proceed.

Note there is an alternative I can easily try and that is to instantiate a ShuffleManager wrapper, which would remove the change to the SparkEnv (we would instantiate the wrapper instead of the actual impl). We could then set the impl on this wrapper at a later time, when jars are localized and plugins are loaded. This felt a bit worse than the approach I have in this PR, but I am happy to hear opinions.

Thanks again!!

File: core/src/main/scala/org/apache/spark/SparkEnv.scala [hunk 6]
```diff
@@ -402,7 +405,7 @@ object SparkEnv extends Logging {
             None
           }, blockManagerInfo,
           mapOutputTracker.asInstanceOf[MapOutputTrackerMaster],
-          shuffleManager,
+          shuffleBlockGetterFn,
```
Comment (beliefer) [inline core/src/main/scala/org/apache/spark/SparkEnv.scala:408]: Why not define `shuffleBlockGetterFn` in `BlockManagerMasterEndpoint`?

**Comment by mridulm** [comment]
While I am not opposed to a way to create a short name for shuffle manager, if it results in nontrivial changes to Spark, I am not very inclined towards it.
IMO this should be something that is better handled in context of SPARK-25299, which unfortunately is partly done.

File: core/src/main/scala/org/apache/spark/SparkEnv.scala (hunk 6 above)
Comment (abellina) [reply core/src/main/scala/org/apache/spark/SparkEnv.scala:410]: I see this being an issue in tests where the `SparkEnv` would not be set, so now I'd have to make sure that the env is set and cleared in the tests. That said, if you feel strongly about this, I can look at this more.

**Comment by tgravescs** [comment]
I agree that ideally we would finish SPARK-25299, I don't see that happening anytime soon.  I also don't think it covers the case of people replacing the entire ShuffleManager vs just the storage piece.   ShuffleManager API isn't public either but we have multiple implementations doing that now (Ubers RSS, project Gluten, Spark Rapids, I thought Cosco was although its not open source, etc).  
One note is that issue SPARK-25299 had a sub issue that was going to use the SparkPlugin for configuration https://issues.apache.org/jira/browse/SPARK-30033/https://github.com/apache/spark/pull/26670 and had a pr that mentions the weird interaction with initialization and it works around it in a different way.

Overall while there are a bunch of changes here most of it is just moving initialization stuff around that shouldn't impact anything else. The one thing that is user impacting is the SparkEnv api change, which if we only do with 4.0 shouldn't be a big deal, unless there is some usage I'm not aware of.  @mridulm  Is there a specific part you are concerned with?

**Comment by abellina** [comment]
@tgravescs @mridulm @beliefer I made a small tweak where the `executorEnvs` map in the `SparkContext` is populated with the configuration prefix `spark.executorEnv.*` after the driver plugin is instantiated (see the last two commits).

**Comment by mridulm** [comment]
@tgravescs The SparkEnv related change is what gave me pause ... I am less concerned about the Executor side of things

**Review by mridulm — COMMENTED** [review]
I am yet to test my comments, but you can reference a version with the changes [here](https://github.com/apache/spark/compare/master...mridulm:spark:SPARK-45762-alternative-proposal)
(SparkSubmitSuite passes).

File: core/src/main/scala/org/apache/spark/SparkEnv.scala [hunk 7]
```diff
@@ -277,6 +287,11 @@ object SparkEnv extends Logging {
       hostname, numCores, ioEncryptionKey, isLocal)
   }
 
+  private def shuffleBlockGetterFn(shuffleId: Int, mapId: Long): Seq[BlockId] = {
+    val env = SparkEnv.get
+    env.shuffleManager.shuffleBlockResolver.getBlocksForShuffle(shuffleId, mapId)
+  }
+
```
[the remark this answers is not in the document]
Comment (mridulm) [reply core/src/main/scala/org/apache/spark/SparkEnv.scala:294]: Looking at this later, preserving this is mainly to minimize test code.

**Comment by mridulm** [comment]
@abellina, given [SPARK-45792](https://issues.apache.org/jira/browse/SPARK-45792) is an SPIP, can you please surface in spark-dev@ and initiate a discussion on it ? I dont remember seeing it there.

**Comment by abellina** [comment]
@mridulm thanks for the comments. I think they mostly work and I’m testing it at the moment. I’ll push something either tonight or tomorrow am.

On the SPIP, I did send it to dev as a DISCUSS thread on November 4. I would like to get input on that as well.

**Comment by mridulm** [comment]
> On the SPIP, I did send it to dev as a DISCUSS thread on November 4. I would like to get input on that as well.

You are right, I do [see it here](https://lists.apache.org/thread/m5pv7vzg3109ts706b13t1c9p5zys2rq) - looks like I missed it :-(

**Comment by abellina** [comment]
@mridulm updated. Thanks for your comments and patch. I took most of it verbatim and cleaned up a couple of things.

**Review by mridulm — COMMENTED** [review]
Can you please check the [latest diffs](https://github.com/apache/spark/compare/master...mridulm:spark:SPARK-45762-alternative-proposal) - there were a few things I had initally missed (for example, in `initiailzeShuffleManager`, etc).

Given this is based on what I proposed, would be better is @tgravescs reviews it once you have had a chance to update the PR @abellina !

**Comment by abellina** [comment]
Thanks @mridulm, yes the commits make sense, it brings back the late initialization in the driver. I tested the change, the main difference from your patch @mridulm is I had to still get the shuffle manage class names using the method we added to the `ShuffleManager` object here https://github.com/apache/spark/pull/43627/files#diff-42a673b8fa5f2b999371dc97a5de7ebd2c2ec19447353d39efb7e8ebc012fe32R592, because the `shuffleManager` is not set yet at this point.

@tgravescs fyi

File: core/src/main/scala/org/apache/spark/SparkEnv.scala [hunk 8]
```diff
@@ -71,6 +70,12 @@ class SparkEnv (
     val outputCommitCoordinator: OutputCommitCoordinator,
     val conf: SparkConf) extends Logging {
 
+  // We initialize the ShuffleManager later in SparkContext, and Executor, to allow
```
Comment (tgravescs) [inline core/src/main/scala/org/apache/spark/SparkEnv.scala:73]:
```suggestion
  // We initialize the ShuffleManager later in SparkContext and Executor to allow
```

File: core/src/main/scala/org/apache/spark/SparkEnv.scala [hunk 9]
```diff
@@ -415,6 +418,11 @@ object SparkEnv extends Logging {
         advertiseAddress, blockManagerPort, numUsableCores, blockManagerMaster.driverEndpoint)
 
     // NB: blockManager is not valid until initialize() is called later.
+    //     SPARK-45762 introduces a change where the ShuffleManager is initialized later
+    //     in the SparkContext and Executor, to allow for custom ShuffleManagers defined
+    //     in user jars. In the executor, the BlockManager uses a lazy val to obtain the
+    //     shuffleManager from the SparkEnv. In the driver, the SparkEnv's shuffleManager
```
Comment (tgravescs) [inline core/src/main/scala/org/apache/spark/SparkEnv.scala:424]: I think this comment it no longer true. Driver SparkEnv shufflemanager is created after the plugin initialized.

Comment (abellina) [reply core/src/main/scala/org/apache/spark/SparkEnv.scala]: Thanks @tgravescs. Handled both comments here: https://github.com/apache/spark/pull/43627/commits/6d002a361ac2c1dfad48ee530766c9b0a605696f

**Comment by abellina** [comment]
There were some CI failures around missing dependencies in the documentation build (all tests are passing otherwise). So I have upmerged. I also tweaked a couple of comments here: https://github.com/apache/spark/pull/43627/commits/5480faa160bb704f9eb85d51e3fa64d15305268e

**Review by tgravescs — APPROVED** [review]
This looks good to me.

**Review by mridulm — APPROVED** [review]

**Comment by mridulm** [comment]
Merged to master.
Thanks for working on this @abellina !
Thanks for the reviews @tgravescs, @beliefer  :-)

**Comment by tgravescs** [comment]
Thanks @mridulm @abellina

**Comment by abellina** [comment]
Thanks @mridulm @tgravescs and @beliefer for the reviews and rework!

**Review by dongjoon-hyun — COMMENTED** [review]
Hi, @abellina , @mridulm , @tgravescs, @beliefer .

Although this is a developer API, this is a documented one. Do you think we can avoid this breaking change by adding a new constructor instead?

- https://spark.apache.org/docs/3.5.0/api/java/org/apache/spark/SparkEnv.html

**Comment by mridulm** [comment]
@dongjoon-hyun, tt is `@DeveloperApi` from point of view of usage - `SparkEnv` is not expected to be created by users, as some of the constructor parameters are not externally visible (`RpcEnv`, for example, cannot be created as it is `private[spark]`). There have been changes to its constructor in the past as well, after it was marked `@DeveloperApi` - though to be fair, these were a while back.

In general, I am conflicted about trying to preserve compatibility for things which are clearly private to spark - it inhibits the ability for the project to evolve: especially around major version boundaries (though we do have a lot of these instances where we try to maintain compatibility).

Given how long `SparkEnv` has been around, I can see valid case being made for adding a constructor which preserves earlier signature. Thiughts @tgravescs ?

**Comment by dongjoon-hyun** [comment]
@mridulm . Of course, it's legit if it's not easy or there is no other way. Also, we have a similar breaking proposal, #45052 , too. While reviewing that PR, I double-checked this PR briefly.

I'm totally fine if this is inevitable here and there. :)

**Comment by mridulm** [comment]
Thanks for understanding @dongjoon-hyun !

**Pull request closed without merging**